File: L11AtlasianMedium/RankTeamByVotes.java

import java.util.*;

public class RankTeamByVotes {

    public static void main(String[] args) {
        String[] votes = new String[5];
        votes[0] = "ABC";
        votes[1] = "ACB";
        votes[2] = "ABC";
        votes[3] = "ACB";
        votes[4] = "ACB";

        Map<Character, int[]> track = new HashMap<>();
        int length = votes[0].length();

        // Initialize the ranking map
        for (String vote : votes) {
            for (int j = 0; j < length; j++) {
                char ch = vote.charAt(j);
                track.putIfAbsent(ch, new int[length]);
                track.get(ch)[j]++;
            }
        }

        // Create a list of characters (teams)
        List<Character> list = new ArrayList<>(track.keySet());

        // Sort the teams based on ranking
        Collections.sort(list, (a, b) -> {
            for (int j = 0; j < length; j++) {
                if (track.get(a)[j] != track.get(b)[j]) {
                    return track.get(b)[j] - track.get(a)[j]; // Higher votes come first
                }
            }
            return a - b; // If votes are the same, sort lexicographically
        });

        // Build the result string
        StringBuilder sb = new StringBuilder();
        for (char ch : list) {
            sb.append(ch);
        }

        // Print the result
        System.out.println(sb.toString());
    }
}

File: L11AtlasianMedium/RankTeamByVotes.cpp
#include <bits/stdc++.h>
using namespace std;

unordered_map<char, vector<int> > track;

// Custom comparator function to sort candidates based on their ranks
bool cmp(char a, char b)
{
    int n = track[a].size();
    for (int j = 0; j < n; j++)
    {
        if (track[a][j] > track[b][j])
            return true;
        else if (track[a][j] < track[b][j])
            return false;
    }
    return a < b; // If ranks are the same, sort lexicographically
}

int main()
{
    vector<string> votes = {"ABC", "ACB", "ABC", "ACB", "ACB"};
    int n = votes[0].size();

    // Initialize the tracking map with zeros
    for (char c : votes[0])
    {
        track[c] = vector<int>(n, 0);
    }

    // Count the votes for each position
    for (auto &vote : votes)
    {
        for (int j = 0; j < n; j++)
        {
            track[vote[j]][j]++;
        }
    }

    // Create a string containing all unique candidates
    string ans = votes[0];

    // Sort candidates based on the custom comparator
    sort(ans.begin(), ans.end(), cmp);

    // Print the result
    cout << ans << endl;

    return 0;
}
